File: templates/c/advanced_foundations.c
// CodeAps Starter Template: C - ADVANCED FOUNDATIONS\n
// --- EXECUTIVE SUITE: C DYNAMIC STACK ---
// High-performance manual memory management example

#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *array;
    int top;
    int capacity;
} Stack;

Stack* createStack(int cap) {
    Stack *s = (Stack*)malloc(sizeof(Stack));
    s->array = (int*)malloc(cap * sizeof(int));
    s->top = -1;
    s->capacity = cap;
    return s;
}

void push(Stack *s, int val) {
    if (s->top < s->capacity - 1) {
        s->array[++(s->top)] = val;
        printf("Pushed: %d\n", val);
    }
}

int pop(Stack *s) {
    if (s->top >= 0) {
        return s->array[(s->top)--];
    }
    return -1;
}

int main() {
    Stack *myStack = createStack(10);
    push(myStack, 42);
    push(myStack, 1337);
    
    printf("Popped: %d\n", pop(myStack));
    return 0;
}

File: templates/rust/advanced_foundations.rs
// CodeAps Starter Template: RUST - ADVANCED FOUNDATIONS\n
// --- EXECUTIVE SUITE: GENERIC STACK ---
// Type-safe modular stack with proper Error Handling

#[derive(Debug)]
pub struct Stack<T> {
    elements: Vec<T>,
    capacity: usize,
}

impl<T> Stack<T> {
    pub fn new(capacity: usize) -> Self {
        Stack {
            elements: Vec::with_capacity(capacity),
            capacity,
        }
    }

    pub fn push(&mut self, item: T) -> Result<(), &str> {
        if self.elements.len() < self.capacity {
            self.elements.push(item);
            Ok(())
        } else {
            Err("Buffer Overflow")
        }
    }

    pub fn pop(&mut self) -> Option<T> {
        self.elements.pop()
    }

    pub fn peek(&self) -> Option<&T> {
        self.elements.last()
    }
}

fn main() {
    let mut brain = Stack::new(3);
    
    match brain.push("Neural Link Alpha") {
        Ok(_) => println!("Synapse active."),
        Err(e) => println!("Warning: {}", e),
    }

    if let Some(data) = brain.peek() {
        println!("Buffer Top: {}", data);
    }
}
